File: Frontend/bdd-testing/features/step-definitions/admin_dashboard_steps.js
const { Given, Then, Before, After, AfterAll, setDefaultTimeout } = require('@cucumber/cucumber');
const puppeteer = require('puppeteer');
const assert = require('assert');

setDefaultTimeout(60 * 1000);

let browser;
let page;
let testStats = { passed: 0, failed: 0, skipped: 0 };

Before(async function () {
  console.log('\n  Before');
  browser = await puppeteer.launch({ 
    headless: true,
    args: ['--no-sandbox', '--disable-setuid-sandbox']
  });
  page = await browser.newPage();
  console.log('  ✔ passed\n');
});

After(async function () {
  console.log('\n  After');
  await browser.close();
  console.log('  ✔ passed\n');
});

Given('I am on {string}', { timeout: 10000 }, async function (pageName) {
  const stepText = `Given I am on "${pageName}"`;
  console.log(`  ${stepText}`);
  
  try {
    const response = await page.goto('http://localhost:52330/Frontend/dashboard-admin/index.html', {
      waitUntil: 'domcontentloaded',
      timeout: 8000
    });
    
    console.log(`    Status: ${response.status()}`);
    await page.waitForSelector('body', { timeout: 3000 });
    console.log('    Body content loaded');
    console.log('  ✔ passed');
    testStats.passed++;
    
  } catch (error) {
    console.log(`  ✖ failed — ${error.message}`);
    testStats.failed++;
    throw error;
  }
});

Then('I should see {string}', { timeout: 5000 }, async function (expectedText) {
  const stepText = `Then I should see "${expectedText}"`;
  console.log(`  ${stepText}`);
  
  try {
    await new Promise(resolve => setTimeout(resolve, 500));
  
    const translations = {
      'Total Checked News': 'Total Pengecekan',
      'Total Hoax Detected': 'Hoax Terdeteksi',
      'System Overview': 'Sistem Overview',
      'Machine Learning Version': 'Versi model',
      'Update Model': 'Update Model',
      'Aksi Admin': 'Aksi Admin',
    };
  
    const isVisible = await page.evaluate((text, alt) => {
      const body = (document.body && document.body.innerText) ? document.body.innerText.toLowerCase() : '';
      const t = (text || '').toLowerCase();
      const a = (alt || '').toLowerCase();
      return (t && body.includes(t)) || (a && body.includes(a));
    }, expectedText, translations[expectedText]);
    
    if (isVisible) {
      console.log(`    Found: "${expectedText}"`);
      console.log('  ✔ passed');
      testStats.passed++;
      assert.ok(true);
    } else {
      const currentText = await page.evaluate(() => document.body.innerText);
      console.log(`    Not found on page`);
      console.log(`    Page content: ${currentText.slice(0, 200)}...`);
      console.log('  ✖ failed');
      testStats.failed++;
      assert.fail(`Expected to find "${expectedText}" on the page`);
    }
  } catch (error) {
    console.log(`  ✖ failed — ${error.message}`);
    testStats.failed++;
    throw error;
  }
});

AfterAll(async function () {
  console.log('\n\n========== TEST SUMMARY ==========');
  console.log(`✔ passed:  ${testStats.passed}`);
  console.log(`✖ failed:  ${testStats.failed}`);
  console.log(`⊘ skipped: ${testStats.skipped}`);
  console.log(`━ total:   ${testStats.passed + testStats.failed + testStats.skipped}`);
  console.log('==================================\n');
});

File: Frontend/bdd-testing/features/steps/admin_dashboard_steps.py
from behave import given, then
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC

@given('I am on "Admin Homepage"')
def step_on_admin_homepage(context):
    context.browser.get("http://localhost:3001/")

@then('I should see "Total Pengecekan"')
def step_see_total_pengecekan(context):
    WebDriverWait(context.browser, 10).until(EC.presence_of_element_located((By.XPATH, "//*[contains(text(),'Total Pengecekan')]")))
    assert "Total Pengecekan" in context.browser.page_source

@then('I should see "Hoax Terdeteksi"')
def step_see_hoax_terdeteksi(context):
    assert "Hoax Terdeteksi" in context.browser.page_source

@then('I should see "Sistem Overview"')
def step_see_sistem_overview(context):
    assert "Sistem Overview" in context.browser.page_source

@then('I should see "Update Model"')
def step_see_update_model(context):
    assert "Update Model" in context.browser.page_source

@then('I should see "Aksi Admin"')
def step_see_aksi_admin(context):
    assert "Aksi Admin" in context.browser.page_source